### PR TITLE
[CHNL-29461] Add on/off switch for logging capabilities

### PR DESCRIPTION
**Pull request by ndurell**
## Description

Adds a configurable logging toggle to the Klaviyo Swift SDK, allowing developers to enable or disable all SDK logging at runtime. This addresses [GitHub Issue #475](https://github.com/klaviyo/klaviyo-swift-sdk/issues/475).

### Changes

- **`Sources/KlaviyoCore/Utils/Logger+Ext.swift`**: Added `KlaviyoLogConfig` — a thread-safe singleton with an `isLoggingEnabled` flag. Converted `static let` Logger instances to computed `static var` properties that return `Logger(OSLog.disabled)` when logging is disabled.
- **`Sources/KlaviyoSwift/Utilities/Logger+Ext.swift`**: Converted `static let` loggers to computed properties gated on the toggle.
- **`Sources/KlaviyoForms/Utilities/Logger+Ext.swift`**: Converted `static let` loggers to computed properties gated on the toggle.
- **`Sources/KlaviyoLocation/Utilities/Logger+Ext.swift`**: Converted `static let` logger to computed property gated on the toggle.
- **`Sources/KlaviyoCore/Utils/LoggerClient.swift`**: Gated `LoggerClient.production.error` and `runtimeWarn()` behind the toggle.
- **`Sources/KlaviyoSwift/Klaviyo.swift`**: Added public API — `isLoggingEnabled` (read-only) and `setLoggingEnabled(_:)` (chainable).

### New Public API

```swift
let sdk = KlaviyoSDK()
sdk.isLoggingEnabled        // true (default)
sdk.setLoggingEnabled(false) // silences all SDK logging
sdk.setLoggingEnabled(true)  // resumes logging
```

# Checklist

Authors and Reviewers, please ensure the following:

## General

- [x] This PR has fewer than 500 lines of code changed (not including tests or docs)
- [x] Refactors must be submitted in separate PRs from functional or logic changes
- [ ] If the reviewer isn't entirely familiar with the context of the change, an SME has been tagged

## Tests

- [x] Each functional change has corresponding test cases that have been added/modified
- [x] Edge cases have been considered and tested
- [x] Mocked data makes sense with what is being tested

## Test Plan

- [x] The author has provided a test plan (in the `Test Plan` section below, including evidence that it was carried out)
- [ ] Reviewer(s) have reviewed the test plan and have a high level of confidence in the change (Please add a comment to the PR indicating your level of confidence on a scale of 1-5, where 5 is very confident, and sign-off in your review). Low confidence ratings should prompt a discussion with the author.
- [ ] If the PR involves changes to event content, process flows or side effects, the test plan covers thorough testing of this via automated or ondemand tests.

# Risk Impact

- **Downstream impacts**: None — this is an additive API. Default behavior (logging enabled) is preserved.
- **Performance implications**: Negligible. The `NSLock` + `Bool` check is sub-microsecond. Logger instances are now computed properties instead of stored, but `Logger(OSLog.disabled)` is optimized by the OS.
- **Environment-specific flags**: None.
- **Forward compatible**: Yes — new API only, no data structure changes.
- **Backwards compatible**: Yes — default is `isLoggingEnabled = true`, preserving current behavior. Safe to revert.

# Test Plan

Automated implementation via AI Agent for CHNL-29461.

Changes implemented:
- Added `KlaviyoLogConfig` thread-safe singleton in `Sources/KlaviyoCore/Utils/Logger+Ext.swift`
- Converted all module Logger instances to toggle-aware computed properties
- Gated `LoggerClient.production.error` and `runtimeWarn()` on the toggle
- Added `setLoggingEnabled(_:)` and `isLoggingEnabled` to `KlaviyoSDK`

Testing performed:
- **7 new unit tests** in `Tests/KlaviyoCoreTests/KlaviyoLogConfigTests.swift`:
  - Default enabled state
  - Disable/re-enable toggle
  - Logger instances return real/disabled loggers correctly
  - Thread safety under concurrent read/write (1000 iterations)
  - Concurrent toggle + logger access (500 iterations)
- **4 new unit tests** in `Tests/KlaviyoSwiftTests/KlaviyoSDKTests.swift`:
  - `isLoggingEnabled` defaults to true
  - `setLoggingEnabled(false)` disables
  - Re-enabling restores logging
  - `setLoggingEnabled` is chainable
- All 87 KlaviyoCoreTests pass ✅
- All 28 KlaviyoSwiftTests pass ✅
- Full package build succeeds ✅

**Note**: Pre-commit hooks (SwiftLint/SwiftFormat) could not be run locally due to environment setup issues (SwiftLint build failure in pre-commit sandbox). CI should run these checks.

## Canary / Manual Verification

This change is fully covered by automated tests. Manual verification can be done by:
1. Calling `KlaviyoSDK().setLoggingEnabled(false)` in a test app
2. Triggering SDK operations (API calls, events, etc.)
3. Verifying no log output appears in Console.app
4. Calling `KlaviyoSDK().setLoggingEnabled(true)` and verifying logs resume

# Deploy Plan

Standard deployment — merge when CI passes and reviews are complete.

If a rollback is needed: revert this PR and the previous behavior will be restored.

# Post-Deploy Monitoring

1. Monitor application logs for any new errors related to the logging toggle
2. Verify SDK initialization and logging work as expected through spot-checking
3. No new metrics dashboards needed — this is a client-side logging control

Closes CHNL-29461

🤖 Generated with [VK Conductor](https://github.com/klaviyo/ai-assist)

## Release Notes

* **New Features**
  * Added dynamic logging control to the SDK. You can now read the current `isLoggingEnabled` state and toggle logging at runtime with `setLoggingEnabled(_:)` (supports chaining).
* **Bug Fixes**
  * Logging output is now suppressed consistently when logging is disabled, including diagnostic warnings.
* **Documentation**
  * Updated the README with a new “Logging” section and usage examples.
* **Tests**
  * Added/expanded test coverage for logging toggle behavior and concurrent safety.

### CODE REVIEW

File: Sources/KlaviyoCore/Utils/Logger+Ext.swift [hunk 1]
```diff
@@ -5,8 +5,40 @@
 //  Created by Andrew Balmer on 7/25/25.
 //
 
+import Foundation
 import OSLog
 
+// MARK: - Logging Configuration
+
+/// Thread-safe global logging toggle for the Klaviyo SDK.
+///
```
Comment (ndurell) [inline Sources/KlaviyoCore/Utils/Logger+Ext.swift]: **Consider using `package` access level instead of `public`.**

Since the package uses `swift-tools-version: 5.9`, you can use the `package` access level here. This would make `KlaviyoLogConfig` accessible from all modules within the package (KlaviyoCore, KlaviyoSwift, KlaviyoForms, KlaviyoLocation) but **not** from external SDK consumers.

Currently, with `public` access, SDK consumers can bypass the intended `KlaviyoSDK.setLoggingEnabled(_:)` API and directly call `KlaviyoLogConfig.shared.isLoggingEnabled = false`. Using `package` access would enforce the designed API surface:

```swift
package final class KlaviyoLogConfig: @unchecked Sendable {
    package static let shared = KlaviyoLogConfig()

    private let lock = NSLock()
    private var _isLoggingEnabled: Bool = true

    private init() {}

    package var isLoggingEnabled: Bool {
        // ...
    }
}
```

This ensures consumers use `KlaviyoSDK().setLoggingEnabled(_:)` as intended.

File: Sources/KlaviyoForms/Utilities/Logger+Ext.swift [hunk 2]
```diff
@@ -5,6 +5,7 @@
 //  Created by Andrew Balmer on 1/28/25.
 //
 
+import KlaviyoCore
```
Comment (ndurell) [inline Sources/KlaviyoForms/Utilities/Logger+Ext.swift]: **Missing explicit dependency in Package.swift.**

This file now imports `KlaviyoCore` directly, but the `KlaviyoForms` target in `Package.swift` only declares `KlaviyoSwift` as a dependency:

```swift
.target(
    name: "KlaviyoForms",
    dependencies: ["KlaviyoSwift"],  // ← no explicit KlaviyoCore
    ...
)
```

This works today because `KlaviyoSwift` transitively depends on `KlaviyoCore`, but relying on transitive imports is fragile. If `KlaviyoSwift` ever refactored its dependency graph, this would silently break.

Please add `"KlaviyoCore"` as an explicit dependency:

```swift
.target(
    name: "KlaviyoForms",
    dependencies: ["KlaviyoSwift", "KlaviyoCore"],
    ...
)
```

**Review by ndurell — COMMENTED** [review]
## AI Review — Round 1

Posted 2 inline comment(s) on specific lines. Please address each one.

**Summary of findings:**

1. **Package.swift: Missing explicit KlaviyoCore dependency for KlaviyoForms** — The PR adds `import KlaviyoCore` to KlaviyoForms but doesn't update Package.swift to declare the direct dependency. It compiles via transitive dependency through KlaviyoSwift, but this should be made explicit.

2. **KlaviyoLogConfig: Consider `package` access level** — Since the package uses swift-tools-version 5.9, `KlaviyoLogConfig` can use `package` access instead of `public` to prevent SDK consumers from bypassing the intended `KlaviyoSDK.setLoggingEnabled(_:)` API.

**What looks good:**
- Thread-safe implementation with NSLock + `@unchecked Sendable` is correct
- All four module Logger extensions consistently updated to computed properties
- LoggerClient.production.error and runtimeWarn() properly gated
- Chainable API follows existing KlaviyoSDK patterns exactly
- Test coverage is thorough (thread safety, toggle behavior, SDK integration)
- Default behavior preserved (logging enabled)
- No breaking changes

---
*Posted by VK Conductor PR Review agent. Addressing this feedback will trigger another review cycle.*

**Review by ndurell — COMMENTED** [review]
## AI Review — Round 3

LGTM — no new issues found. All prior feedback (Round 1 AI review, Bugbot rounds 1 & 2) has been properly addressed across commits da46dc7 and 1b32912:

- **Package.swift**: KlaviyoForms declares explicit KlaviyoCore dependency ✓
- **KlaviyoLogConfig**: Uses `package` access level ✓
- **EventBuffer.swift**: Logger converted to toggle-aware computed property ✓
- **KlaviyoSDKTests**: tearDown resets logging state ✓
- **Vendor runtimeWarn**: Gated by `KlaviyoLogConfig.shared.isLoggingEnabled` ✓

All acceptance criteria verified:
1. `setLoggingEnabled(_:)` — present, chainable, `@discardableResult`
2. `isLoggingEnabled` — present, read-only
3. Default enabled — `_isLoggingEnabled = true`
4. All loggers silenced when disabled — KlaviyoCore, KlaviyoSwift, KlaviyoForms, KlaviyoLocation, LoggerClient, both runtimeWarn copies
5. Re-enable works immediately — computed properties
6. Thread-safe — NSLock + `@unchecked Sendable` + concurrent tests

Proceeding to CI.

---
*Posted by VK Conductor PR Review agent.*

**Review by ndurell — COMMENTED** [review]
## AI Review — Round 4

LGTM — no new issues found. Max review rounds (3) reached — proceeding to CI.

All prior feedback has been properly addressed across commits da46dc7, 1b32912, e384600, and 952fc83. Implementation verified against all 6 acceptance criteria.

---
*Posted by VK Conductor PR Review agent.*

**Review by ab1470 — APPROVED** [review]
This looks great to me! Thanks for picking this up

**Comment by ajaysubra** [comment]
@coderabbitai review

File: Sources/KlaviyoLocation/Utilities/Logger+Ext.swift [hunk 3]
```diff
@@ -5,12 +5,15 @@
 //  Created by Andrew Balmer on 10/8/24.
 //
 
+import KlaviyoCore
 import OSLog
 
 @available(iOS 14.0, *)
 extension Logger {
     private static var subsystem = Bundle.main.bundleIdentifier ?? ""
```
[the remark this answers is not in the document]
Comment (ndurell) [reply Sources/KlaviyoLocation/Utilities/Logger+Ext.swift]: Addressed in 5970121 — the KlaviyoLocation subsystem is now `"com.klaviyo.klaviyo-swift-sdk.klaviyoLocation"`, consistent with the other modules.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Sources/KlaviyoSwift/Vendor/ComposableArchitecture/Misc.swift [hunk 4]
```diff
@@ -46,7 +47,6 @@ final class Box<Wrapped> {
   }
 }
 
-@_transparent
```
[the remark this answers is not in the document]
Comment (ndurell) [reply Sources/KlaviyoSwift/Vendor/ComposableArchitecture/Misc.swift:49]: Intentional — `@_transparent` functions can't reference `package`-scoped symbols like `KlaviyoLogConfig` (compiler error). `@inline(__always)` is retained; see the fuller rationale in https://github.com/klaviyo/klaviyo-swift-sdk/pull/523#issuecomment-4042313141. This also matches the KlaviyoCore `runtimeWarn`, which never had `@_transparent`.

---
_Generated by [Claude Code](https://claude.ai/code)_
